File: bindings/swift/Tests/ZtokTests/FuzzTests.swift
// FuzzTests — PRNG-driven round-trip fuzz harness for the Swift
// binding. Mirrors fuzz/encode_decode.zig and the corresponding
// Python/Node/Ruby/Rust/Java suites: a deterministic LCG mutates the
// byte input each iteration, encode-then-decode must round-trip
// exactly.
//
// The pipeline is byte_id (identity normalizer + identity pre-tok +
// byte_id model + concat decoder); by construction each input byte
// maps to one id and decoding concatenates them back, so for any
// byte sequence the invariant decode(encode(x)) == x must hold.
//
// Seed defaults to 0xFEEDB0B (matching the other bindings) — failures
// at a given iteration are cross-language reproducible. Override via
// `FUZZ_SEED` (decimal or `0x...` hex) and `ZTOK_FUZZ_ITERS`
// environment variables for nightly runs.

import XCTest

@testable import Ztok

final class FuzzTests: XCTestCase {
    private let defaultSeed: UInt64 = 0xFEEDB0B
    private let defaultIterations: Int = 1000
    private let maxLen: Int = 256

    override func setUpWithError() throws {
        try XCTSkipUnless(TestSupport.libztokAvailable,
                          "libztok not loadable (set LD_LIBRARY_PATH or build with `zig build`)")
    }

    func testByteIdRoundTripFuzz() throws {
        let env = ProcessInfo.processInfo.environment
        let seed = env["FUZZ_SEED"].flatMap(parseU64) ?? defaultSeed
        let iters = env["ZTOK_FUZZ_ITERS"].flatMap { Int($0) } ?? defaultIterations

        let pipe = try Pipeline.byteId()
        defer { pipe.close() }

        var rng = SplitMix64(seed: seed)
        var failures: [(Int, [UInt8], [UInt8])] = []

        for i in 0..<iters {
            let len = Int(rng.next() % UInt64(maxLen + 1))
            var data = [UInt8](repeating: 0, count: len)
            for j in 0..<len {
                data[j] = UInt8(rng.next() & 0xFF)
            }

            let ids = try pipe.encodeBytes(data)
            XCTAssertEqual(ids.count, data.count,
                           "iter \(i): byte_id produced \(ids.count) ids for \(data.count) bytes")
            let roundTrip = try pipe.decodeBytes(ids)
            if roundTrip != data {
                failures.append((i, data, roundTrip))
                if failures.count >= 5 { break }
            }
        }

        if !failures.isEmpty {
            var msg = "byte_id round-trip mismatches:\n"
            for (i, inBytes, outBytes) in failures {
                msg += "  iter \(i): in=\(inBytes) out=\(outBytes)\n"
            }
            XCTFail(msg)
        }
    }
}

/// Parse a u64 from decimal or `0x...` / `0X...` hex.
private func parseU64(_ raw: String) -> UInt64? {
    if raw.hasPrefix("0x") || raw.hasPrefix("0X") {
        return UInt64(raw.dropFirst(2), radix: 16)
    }
    return UInt64(raw)
}

/// Deterministic 64-bit PRNG (SplitMix64). Same shape across
/// bindings — picking any fixed PRNG is fine for fuzz coverage, and
/// SplitMix64 is small enough to inline here without pulling in
/// `swift-numerics`.
struct SplitMix64 {
    private var state: UInt64

    init(seed: UInt64) { self.state = seed }

    mutating func next() -> UInt64 {
        state &+= 0x9E3779B97F4A7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58476D1CE4E5B9
        z = (z ^ (z >> 27)) &* 0x94D049BB133111EB
        return z ^ (z >> 31)
    }
}

File: bindings/dotnet/Ztok.Tests/FuzzTests.cs
// FuzzTests.cs — PRNG-driven round-trip fuzz harness.
//
// Mirrors fuzz/encode_decode.zig and the Python/Node/Ruby/Rust fuzz
// tests in shape: a deterministic PRNG mutates the byte input each
// iteration, encode-then-decode must round-trip exactly. The pipeline
// is byte_id (each input byte maps to one id), so for any byte sequence
// the invariant decode(encode(x)) == x must hold.
//
// Seed defaults to 0xFEEDB0B to match the Python/Ruby/Node/Rust
// harnesses — failures at a given iteration are cross-language
// reproducible. Override via FUZZ_SEED (hex or decimal) and
// ZTOK_FUZZ_ITERS for nightly runs.

using System;
using System.Collections.Generic;
using System.Globalization;
using Xunit;
using Ztok;

namespace Ztok.Tests;

public class FuzzTests
{
    // System.Random with a seeded int gives a deterministic stream. We
    // use the seeded int directly (no Random.Shared) so reproducibility
    // is unconditional.
    private const int DefaultSeed = unchecked((int)0x0FEEDB0B);
    private const int DefaultIterations = 1000;
    private const int MaxLen = 256;

    private static int EnvInt(string name, int defaultValue)
    {
        var raw = Environment.GetEnvironmentVariable(name);
        if (string.IsNullOrEmpty(raw)) return defaultValue;
        // Accept hex like "0xFEEDB0B" and decimal.
        if (raw.StartsWith("0x", StringComparison.OrdinalIgnoreCase) ||
            raw.StartsWith("0X", StringComparison.OrdinalIgnoreCase))
        {
            if (int.TryParse(raw.AsSpan(2), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var hex))
                return hex;
            return defaultValue;
        }
        return int.TryParse(raw, out var dec) ? dec : defaultValue;
    }

    [Fact]
    public void ByteIdRoundtripFuzz1000Iterations()
    {
        if (!SmokeTests.LibAvailable())
        {
            Console.Error.WriteLine("libztok not available; skipping fuzz harness.");
            return;
        }

        int seed = EnvInt("FUZZ_SEED", DefaultSeed);
        int iters = EnvInt("ZTOK_FUZZ_ITERS", DefaultIterations);
        var rng = new Random(seed);

        var failures = new List<(int Iter, byte[] Input, byte[] Output)>();

        using var pipe = Pipeline.ByteId();
        for (int i = 0; i < iters; i++)
        {
            int n = rng.Next(0, MaxLen + 1);
            var data = new byte[n];
            if (n > 0) rng.NextBytes(data);

            var ids = pipe.EncodeBytes(data);
            // byte_id maps 1:1: id count must equal input byte count.
            Assert.Equal(data.Length, ids.Length);

            var roundtrip = pipe.DecodeBytes(ids);
            if (!roundtrip.AsSpan().SequenceEqual(data))
            {
                failures.Add((i, data, roundtrip));
                if (failures.Count >= 5) break;
            }
        }

        if (failures.Count > 0)
        {
            var msg = "byte_id round-trip mismatches:\n";
            foreach (var (iter, inp, outp) in failures)
            {
                msg += $"  iter {iter}: in={Convert.ToHexString(inp)} out={Convert.ToHexString(outp)}\n";
            }
            Assert.Fail(msg);
        }
    }
}
